pkgs/applications/misc/koreader/src.nix

```diff
@@ -0,0 +1,243 @@
+    # that commit.  For cloned-by-tag repositories we create a tag
+    # with the expected name; for cloned-by-hash repositories we
+    # create a tag named ${placeholder-git-tag}, which is
+    # substitute(d)InPlace for the hash in the CMakeLists.txt during the
+    # `postUnpack` phase.
+    mapAttrsToConcatStringSep "\n"
+      (name: attrs: lib.optionalString (attrs.isGit) ''
+        git -C ${attrs.dest} init
+        git -C ${attrs.dest} config --local user.email "you@example.com"
+        git -C ${attrs.dest} config --local user.name  "Your Name"
+        git -C ${attrs.dest} add -A
+        git -C ${attrs.dest} commit --quiet -m postPatch || true
+        git -C ${attrs.dest} tag -f ${attrs.tag or placeholder-git-tag}
+        '')
+      vendored;
+
+  nativeBuildInputs = [
+    git
+    cmake
+    autoconf
+    automake
+    libtool
+    pkgconfig
+    which
+    perl
+    coreutils
+    luarocks
+    python27
+    python3
+    makeWrapper
+  ];
+
+  buildInputs = [
+    gtk3-x11
+    SDL2
+    glib
+    luaPackages.luajson
+  ];
+
+  dontUseCmakeConfigure = true;
+
+  # nativeBuildInputs=[cmake] enables this; we must re-disable it
+  enableParallelBuilding = false;
+
+  preBuild = ''
+    MACHINE=$(${stdenv.cc}/bin/cc -dumpmachine)
+    ${luarocks}/bin/luarocks install --tree=base/build/$MACHINE/rocks ${buildLuaRock "lpeg"}/*.rock
+    ${luarocks}/bin/luarocks install --tree=base/build/$MACHINE/rocks ${buildLuaRock "luajson"}/*.rock
+    '';
+
+  makeFlags = [
+    "PARALLEL_JOBS=$NIX_BUILD_CORES"
+    "IS_RELEASE=1"
+  ];
+
+  # koreader does not have a separate `make install` target
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out
+
+    mkdir -p $out/share/{applications,pixmaps,man/man1}
+    cp platform/debian/koreader.desktop $out/share/applications/
+    cp platform/debian/koreader.1       $out/share/man/man1/
+    cp resources/koreader.png           $out/share/pixmaps/
+
+    mkdir -p $out/bin
+    cp platform/debian/koreader.sh $out/bin/koreader
+    patchShebangs $out/bin/koreader
+
+    mkdir -p $out/lib
+    cp -rL koreader-emulator-$(${stdenv.cc}/bin/cc -dumpmachine)/koreader $out/lib/koreader
+    ${patchelf}/bin/patchelf --add-rpath $out/lib/koreader/libs/ $out/lib/koreader/libs/*.so*
+    ${patchelf}/bin/patchelf --allowed-rpath-prefixes /nix/store --shrink-rpath $out/lib/koreader/libs/*.so*
+
+    wrapProgram $out/bin/koreader --prefix LD_LIBRARY_PATH : ${
+      lib.makeLibraryPath [ gtk3-x11 SDL2 glib ]
+    }
+
+    # ideally "git describe HEAD | xargs git show -s --format=format:"%cd" --date=short"
+    echo "v${version}" > $out/lib/koreader/git-rev
+
+    runHook postInstall
+    '';
+
+  meta = with lib; {
+    homepage = "https://github.com/koreader/koreader";
+    description = "An ebook reader supporting PDF, DjVu, EPUB, FB2 and other formats";
+    license = licenses.agpl3Only;
+    maintainers = with maintainers; [ contrun neonfuz amjoseph ];
+    sourceProvenance = [ ];
```

Comment:
Is this necessary? Usually empty lists are a "default value".

Reply:
I swear I remember something about `sourceProvenance` being missing meaning "nobody has checked yet" and `[ ]` means "somebody has checked and did not find any not-built-from-source components.

However I'm now unable to find that discussion.  And, apparently it got lost in the RFC/merge process.  What's in the manual right now leaves no way to distinguish between "nobody has checked" and "somebody checked and it's all built from source".

Grrr.

I will change this to `with lib.sourceTypes; [ fromSource ];`.

Reply:
815ea9ac048